spring-integration-core/src/main/java/org/springframework/integration/dispatcher/BroadcastingDispatcher.java

```diff
@@ -158,13 +160,23 @@ public boolean dispatch(Message<?> message) {
 			throw new MessageDispatchingException(message, "Dispatcher has no subscribers");
 		}
 		int sequenceSize = handlers.size();
+		Message<?> messageToSend = message;
+		UUID sequenceId = null;
+		if (this.applySequence) {
+			sequenceId = message.getHeaders().getId();
+		}
 		for (MessageHandler handler : handlers) {
-			Message<?> messageToSend = message;
 			if (this.applySequence) {
-				messageToSend = getMessageBuilderFactory()
-						.fromMessage(message)
-						.pushSequenceDetails(message.getHeaders().getId(), sequenceNumber++, sequenceSize)
-						.build();
+				if (message instanceof ThreadStatePropagationChannelInterceptor.MessageWithThreadState) {
```

Comment:
Rather than tying the dispatcher to interceptors of this type, I think I'd prefer to see a new interface; something like...

```java
/**
  *  A Message<?> implementation that needs special handling in MBF.fromMessage().
  */
public interface ClonableMessage {

    MessageBuilder cloneMessage(); // createBuilderForClone() ?

}
```

Then, in `MessageBuilder.fromMessage()`...

```java
if (message instanceof ClonableMessage) {
    return message.cloneMessage();
}
else {
    return new MessageBuilder(...);
}
```

Or, we could leave the clone logic in the dispatcher (using the interface), but I think it really belongs in the builder.